homeassistant/components/aemet/weather_update_coordinator.py

```diff
@@ -0,0 +1,652 @@
+    AEMET_ATTR_DAY,
+    AEMET_ATTR_DIRECTION,
+    AEMET_ATTR_ELABORATED,
+    AEMET_ATTR_FORECAST,
+    AEMET_ATTR_HUMIDITY,
+    AEMET_ATTR_ID,
+    AEMET_ATTR_IDEMA,
+    AEMET_ATTR_MAX,
+    AEMET_ATTR_MIN,
+    AEMET_ATTR_NAME,
+    AEMET_ATTR_PRECIPITATION,
+    AEMET_ATTR_PRECIPITATION_PROBABILITY,
+    AEMET_ATTR_SKY_STATE,
+    AEMET_ATTR_SNOW,
+    AEMET_ATTR_SNOW_PROBABILITY,
+    AEMET_ATTR_SPEED,
+    AEMET_ATTR_STATION_DATE,
+    AEMET_ATTR_STATION_HUMIDITY,
+    AEMET_ATTR_STATION_LOCATION,
+    AEMET_ATTR_STATION_PRESSURE_SEA,
+    AEMET_ATTR_STATION_TEMPERATURE,
+    AEMET_ATTR_STORM_PROBABILITY,
+    AEMET_ATTR_TEMPERATURE,
+    AEMET_ATTR_TEMPERATURE_FEELING,
+    AEMET_ATTR_WIND,
+    AEMET_ATTR_WIND_GUST,
+    ATTR_DATA,
+)
+from aemet_opendata.helpers import (
+    get_forecast_day_value,
+    get_forecast_hour_value,
+    get_forecast_interval_value,
+)
+import async_timeout
+
+from homeassistant.components.weather import (
+    ATTR_FORECAST_CONDITION,
+    ATTR_FORECAST_PRECIPITATION,
+    ATTR_FORECAST_PRECIPITATION_PROBABILITY,
+    ATTR_FORECAST_TEMP,
+    ATTR_FORECAST_TEMP_LOW,
+    ATTR_FORECAST_TIME,
+    ATTR_FORECAST_WIND_BEARING,
+    ATTR_FORECAST_WIND_SPEED,
+)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    ATTR_API_CONDITION,
+    ATTR_API_FORECAST,
+    ATTR_API_HUMIDITY,
+    ATTR_API_PRESSURE,
+    ATTR_API_RAIN,
+    ATTR_API_RAIN_PROB,
+    ATTR_API_SNOW,
+    ATTR_API_SNOW_PROB,
+    ATTR_API_STATION_ID,
+    ATTR_API_STATION_NAME,
+    ATTR_API_STATION_TIMESTAMP,
+    ATTR_API_STORM_PROB,
+    ATTR_API_TEMPERATURE,
+    ATTR_API_TEMPERATURE_FEELING,
+    ATTR_API_TOWN_ID,
+    ATTR_API_TOWN_NAME,
+    ATTR_API_TOWN_TIMESTAMP,
+    ATTR_API_WIND_BEARING,
+    ATTR_API_WIND_MAX_SPEED,
+    ATTR_API_WIND_SPEED,
+    CONDITIONS_MAP,
+    DOMAIN,
+    FORECAST_MODE_DAILY,
+    FORECAST_MODE_HOURLY,
+    WIND_BEARING_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+WEATHER_UPDATE_INTERVAL = timedelta(minutes=10)
+
+
+def format_condition(condition: str) -> str:
+    """Return condition from dict CONDITIONS_MAP."""
+    for key, value in CONDITIONS_MAP.items():
+        if condition in value:
+            return key
+    _LOGGER.error('condition "%s" not found in CONDITIONS_MAP', condition)
+    return condition
+
+
+def format_float(value) -> float:
+    """Try converting string to float."""
+    try:
+        return float(value)
+    except ValueError:
+        return None
+
+
+def format_int(value) -> int:
+    """Try converting string to int."""
+    try:
+        return int(value)
+    except ValueError:
+        return None
+
+
+class TownNotFound(UpdateFailed):
+    """Raised when town is not found."""
+
+
+class WeatherUpdateCoordinator(DataUpdateCoordinator):
+    """Weather data update coordinator."""
+
+    def __init__(self, aemet, latitude, longitude, forecast_mode, hass):
```

Comment:
Always start with `hass` in the signature if `hass` is present in the signature.

Reply:
As above, this is inherited from OpenWeatherMap: https://github.com/home-assistant/core/blob/fd177441b3cfddb22b893b26e91862307bbcc437/homeassistant/components/openweathermap/weather_update_coordinator.py#L54